src/main/java/com/singlestore/jdbc/plugin/codec/VectorCodec.java

```diff
@@ -0,0 +1,119 @@
+import com.singlestore.jdbc.client.ColumnDecoder;
+import com.singlestore.jdbc.client.Context;
+import com.singlestore.jdbc.client.DataType;
+import com.singlestore.jdbc.client.ReadableByteBuf;
+import com.singlestore.jdbc.client.socket.Writer;
+import com.singlestore.jdbc.client.util.MutableInt;
+import com.singlestore.jdbc.plugin.Codec;
+import com.singlestore.jdbc.type.Vector;
+import java.io.IOException;
+import java.sql.SQLDataException;
+import java.sql.SQLException;
+import java.util.Calendar;
+import java.util.EnumSet;
+
+public class VectorCodec implements Codec<Vector> {
+
+  public static final VectorCodec INSTANCE = new VectorCodec();
+
+  private static final EnumSet<DataType> COMPATIBLE_TYPES =
+      EnumSet.of(
+          DataType.FLOAT64_VECTOR,
+          DataType.FLOAT32_VECTOR,
+          DataType.INT64_VECTOR,
+          DataType.INT32_VECTOR,
+          DataType.INT16_VECTOR,
+          DataType.INT8_VECTOR);
+
+  @Override
+  public String className() {
+    return Vector.class.getName();
+  }
+
+  @Override
+  public boolean canDecode(ColumnDecoder column, Class<?> type) {
+    return COMPATIBLE_TYPES.contains(column.getType()) && type.isAssignableFrom(Vector.class);
+  }
+
+  @Override
+  public boolean canEncode(Object value) {
+    return value instanceof Vector;
+  }
+
+  @Override
+  public int getApproximateTextProtocolLength(Object value) throws SQLException {
+    return canEncode(value) ? String.valueOf(value).length() * 3 : -1;
+  }
+
+  @Override
+  public Vector decodeText(
+      ReadableByteBuf buf, MutableInt length, ColumnDecoder column, Calendar cal)
+      throws SQLDataException {
+    if (COMPATIBLE_TYPES.contains(column.getType()) && column.getExtTypeFormat() != null) {
+      byte[] arr = new byte[length.get()];
+      buf.readBytes(arr);
+      int dimensions = Integer.parseInt(column.getExtTypeFormat().split(",")[0]);
+      return Vector.fromData(arr, dimensions, column.getType(), column.isBinary());
+    }
+    buf.skip(length.get());
+    throw new SQLDataException(
+        String.format("Data type %s cannot be decoded as Vector", column.getType()));
+  }
+
+  @Override
+  public Vector decodeBinary(
+      ReadableByteBuf buf, MutableInt length, ColumnDecoder column, Calendar cal)
+      throws SQLDataException {
+    return decodeText(buf, length, column, cal);
+  }
+
+  @Override
+  public void encodeText(
+      Writer encoder, Context context, Object value, Calendar cal, Long maxLength)
+      throws IOException, SQLException {
+    if (!(value instanceof Vector)) {
+      throw new SQLDataException(
+          String.format("Data type %s cannot be encoded as Vector text", value.getClass()));
+    }
+    Vector vector = (Vector) value;
+    if (vector.isBinary()) {
+      byte[] b = vector.getValues();
+      int length = b.length;
+      encoder.writeBytes(BINARY_PREFIX);
+      encoder.writeBytesEscaped(
+          b, maxLength == null ? length : Math.min(length, maxLength.intValue()));
```

Comment:
Should we take BINARY_PREFIX into account when we are limiting length by `maxLength`?

Reply:
maxLength is used to limit only parameter value